File: python/src/pmsec/tools/bundler.py
from __future__ import annotations

import re
from pathlib import Path

from pmsec.util.context import Context
from pmsec.util.io import write_atomic
from pmsec.util.lines import read_key, remove_key, set_key
from pmsec.util.paths import bundle_config_path
from pmsec.util.version import build_preflight

NAME = "bundler"
KEY = "BUNDLE_COOLDOWN"
DOCS = "https://bundler.io/man/bundle-config.1.html"
MIN_BIN = (4, 0, 13)
EXTRAS: list[dict] = []
SEP = ":"

preflight = build_preflight(
    NAME, MIN_BIN,
    "BUNDLE_COOLDOWN is silently ignored. Upgrade bundler (v4.0.13+) to enforce the cooldown.",
)

# bundler stores the cooldown as a plain integer number of days, quoted in the
# YAML config (e.g. `BUNDLE_COOLDOWN: "7"`). Accept it with or without quotes.
_INT_DAYS = re.compile(r'^"?\s*(\d+)\s*"?$')


def path(ctx: Context) -> Path:
    return bundle_config_path(ctx.env, ctx.home)


def _parse_days(value: str | None) -> int | None:
    if value is None:
        return None
    m = _INT_DAYS.match(value)
    return int(m.group(1)) if m else None


def read(ctx: Context) -> dict:
    p = path(ctx)
    raw = p.read_text("utf-8") if p.exists() else ""
    value = read_key(raw, KEY, sep=SEP)
    return {"path": str(p), "configured": value, "days": _parse_days(value), "extras": []}


def write(days: int, ctx: Context) -> dict:
    p = path(ctx)
    raw = p.read_text("utf-8") if p.exists() else ""
    write_atomic(p, set_key(raw, KEY, f'{KEY}: "{days}"', sep=SEP))
    return {"path": str(p)}


def unset(ctx: Context) -> dict:
    p = path(ctx)
    if not p.exists():
        return {"path": str(p), "removed": False}
    before = p.read_text("utf-8")
    after, removed = remove_key(before, KEY, sep=SEP)
    if removed:
        write_atomic(p, after)
    return {"path": str(p), "removed": removed}

File: node/src/tools/bundler.mjs
import { bundleConfigPath } from "../util/paths.mjs";
import { readSafe, writeAtomic } from "../util/io.mjs";
import { readKey, setKey, removeKey } from "../util/lines.mjs";
import { buildPreflight } from "../util/version.mjs";

export const name = "bundler";
export const key = "BUNDLE_COOLDOWN";
export const docs = "https://bundler.io/man/bundle-config.1.html";
export const minBin = [4, 0, 13];
export const extras = [];
const SEP = ":";

export function path(ctx) { return bundleConfigPath(ctx.env, ctx.home); }

export const preflight = buildPreflight(name, minBin,
  "BUNDLE_COOLDOWN is silently ignored. Upgrade bundler (v4.0.13+) to enforce the cooldown.");

// bundler stores the cooldown as a plain integer number of days, quoted in the
// YAML config (e.g. `BUNDLE_COOLDOWN: "7"`). Accept it with or without quotes.
function parseDays(value) {
  if (value === null) return null;
  const m = value.match(/^"?\s*(\d+)\s*"?$/);
  return m ? Number(m[1]) : null;
}

export async function read(ctx) {
  const p = path(ctx);
  const raw = await readSafe(p);
  const value = readKey(raw, key, { sep: SEP });
  return { path: p, configured: value, days: parseDays(value), extras: [] };
}

export async function write(days, ctx) {
  const p = path(ctx);
  const text = setKey(await readSafe(p), key, `${key}: "${days}"`, { sep: SEP });
  await writeAtomic(p, text);
  return { path: p };
}

export async function unset(ctx) {
  const p = path(ctx);
  const before = await readSafe(p);
  const { text: after, removed } = removeKey(before, key, { sep: SEP });
  if (removed) await writeAtomic(p, after);
  return { path: p, removed };
}
